File: 第六次实验/irgen.cpp
/**
 * 中间代码生成器 (Intermediate Code Generator - Quadruples)
 * 编译原理第六次实验
 *
 * 功能：
 *   1. 词法分析：将源代码转换为 Token 流
 *   2. 语法分析：递归下降解析
 *   3. 中间代码生成：生成四元式 (op, arg1, arg2, result)
 *
 * 四元式格式: (运算符, 操作数1, 操作数2, 结果)
 * 示例: (+, x, y, t1) 表示 t1 = x + y
 */

#include <windows.h>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <cstdio>
#include <sstream>

using namespace std;

// ===== Token =====
enum TT {
    T_INT, T_FLOAT, T_VOID, T_IF, T_ELSE, T_WHILE,
    T_RETURN, T_INPUT, T_PRINT,
    T_ID, T_NUM, T_FLT,
    T_PLUS, T_MINUS, T_MUL, T_DIV,
    T_ASG, T_LT, T_GT, T_LE, T_GE, T_EQ, T_NE,
    T_AND, T_OR, T_NOT,
    T_LPA, T_RPA, T_LBR, T_RBR, T_LBK, T_RBK,
    T_SEM, T_COM, T_EOF, T_UNK
};

struct Tok {
    TT type; string val; int ln;
};

// ===== 四元式 =====
struct Quad {
    string op, arg1, arg2, result;
    int id;
};

// ===== 符号表条目 =====
struct Sym {
    string name, type;
    bool isFunc;
};

// ===== 词法分析 =====
vector<Tok> lex(const string& src) {
    vector<Tok> toks;
    size_t p = 0; int ln = 1;
    set<string> kws = {"int","float","void","if","else","while","return","input","print"};

    while (p < src.size()) {
        while (p < src.size() && isspace(src[p])) { if (src[p]=='\n') ln++; p++; }
        if (p >= src.size()) break;

        // 注释
        if (p+1 < src.size() && src[p]=='/' && src[p+1]=='/') {
            while (p < src.size() && src[p]!='\n') p++;
            continue;
        }

        char c = src[p];
        if (isalpha(c) || c=='_') {
            string w;
            while (p < src.size() && (isalnum(src[p])||src[p]=='_')) w += src[p++];
            TT t = T_ID;
            if (w=="int") t=T_INT; else if (w=="float") t=T_FLOAT;
            else if (w=="void") t=T_VOID; else if (w=="if") t=T_IF;
            else if (w=="else") t=T_ELSE; else if (w=="while") t=T_WHILE;
            else if (w=="return") t=T_RETURN; else if (w=="input") t=T_INPUT;
            else if (w=="print") t=T_PRINT;
            toks.push_back({t, w, ln});
        } else if (isdigit(c) || (c=='.' && p+1<src.size() && isdigit(src[p+1]))) {
            string n; bool fl=false;
            if (c=='.') { fl=true; n+=src[p++]; }
            while (p<src.size() && isdigit(src[p])) n+=src[p++];
            if (p<src.size() && src[p]=='.') { fl=true; n+=src[p++]; while(p<src.size()&&isdigit(src[p])) n+=src[p++]; }
            if (p<src.size() && (src[p]=='e'||src[p]=='E')) { fl=true; n+=src[p++]; if(p<src.size()&&(src[p]=='+'||src[p]=='-')) n+=src[p++]; while(p<src.size()&&isdigit(src[p])) n+=src[p++]; }
            toks.push_back({fl?T_FLT:T_NUM, n, ln});
        } else if (p+1 < src.size()) {
            string tw = src.substr(p,2);
            TT t = T_UNK;
            if(tw=="<=")t=T_LE; else if(tw==">=")t=T_GE; else if(tw=="==")t=T_EQ;
            else if(tw=="!=")t=T_NE; else if(tw=="&&")t=T_AND; else if(tw=="||")t=T_OR;
            if (t!=T_UNK) { toks.push_back({t,tw,ln}); p+=2; continue; }
            // single char
            goto single;
        } else {
            single:
            TT t = T_UNK;
            switch(c) {
                case '+':t=T_PLUS;break; case '-':t=T_MINUS;break;
                case '*':t=T_MUL;break; case '/':t=T_DIV;break;
                case '=':t=T_ASG;break; case '<':t=T_LT;break; case '>':t=T_GT;break;
                case '!':t=T_NOT;break;
                case '(':t=T_LPA;break; case ')':t=T_RPA;break;
                case '{':t=T_LBR;break; case '}':t=T_RBR;break;
                case '[':t=T_LBK;break; case ']':t=T_RBK;break;
                case ';':t=T_SEM;break; case ',':t=T_COM;break;
                default:break;
            }
            toks.push_back({t,string(1,c),ln}); p++;
        }
    }
    toks.push_back({T_EOF,"",ln});
    return toks;
}

// ===== 中间代码生成器 =====
class IRGen {
    vector<Tok>& toks;
    size_t pos;
    vector<Quad> quads;
    vector<Sym> syms;
    int tmpCnt;
    int labelCnt;

    Tok& peek() { return toks[pos]; }
    Tok& advance() { return toks[pos++]; }
    bool match(TT t) { if(peek().type==t){advance();return true;} return false; }

    string newTmp() { char b[32]; sprintf(b,"t%d",tmpCnt++); return b; }
    string newLabel() { char b[32]; sprintf(b,"L%d",labelCnt++); return b; }

    void emit(const string& op, const string& a1, const string& a2, const string& res) {
        Quad q; q.op=op; q.arg1=a1; q.arg2=a2; q.result=res; q.id=quads.size();
        quads.push_back(q);
    }

public:
    IRGen(vector<Tok>& t) : toks(t), pos(0), tmpCnt(1), labelCnt(0) {}

    void parse() {
        while (peek().type != T_EOF) {
            if (peek().type==T_INT||peek().type==T_FLOAT||peek().type==T_VOID) {
                parseDeclOrFunc();
            } else {
                parseStmt();
            }
        }
    }

    void parseDeclOrFunc() {
        string typeStr = peek().val; advance();
        if (peek().type == T_ID) {
            string name = peek().val; advance();
            if (peek().type == T_LPA) {
                // 函数声明
                syms.push_back({name, typeStr, true});
                emit("FUNC", name, "", "");
                advance(); // (
                // 参数
                while (peek().type==T_INT||peek().type==T_FLOAT||peek().type==T_VOID) {
                    string pt = peek().val; advance();
                    if (peek().type==T_ID) {
                        string pn = peek().val; advance();
                        syms.push_back({pn, pt, false});
                        emit("PARAM", pn, pt, "");
                    }
                    match(T_SEM) || match(T_COM);
                }
                match(T_RPA);
                if (match(T_LBR)) {
                    while (!match(T_RBR) && peek().type!=T_EOF) parseStmt();
                }
                emit("ENDFUNC", name, "", "");
                match(T_SEM);
            } else {
                // 变量声明
                syms.push_back({name, typeStr, false});
                emit("DECL", name, typeStr, "");
                match(T_SEM);
            }
        }
    }

    void parseStmt() {
        switch (peek().type) {
            case T_INT: case T_FLOAT: {
                string typeStr = peek().val; advance();
                if (peek().type == T_ID) {
                    string name = peek().val; advance();
                    syms.push_back({name, typeStr, false});
                    emit("DECL", name, typeStr, "");
                    if (match(T_ASG)) {
                        string val = parseExpr();
                        emit("=", val, "", name);
                    }
                    match(T_SEM);
                }
                break;
            }
            case T_ID: {
                string name = peek().val; advance();
                if (match(T_ASG)) {
                    string val = parseExpr();
                    emit("=", val, "", name);
                    match(T_SEM);
                } else if (match(T_LPA)) {
                    // 函数调用
                    string args;
                    while (peek().type != T_RPA && peek().type != T_EOF) {
                        string a = parseExpr();
                        if (!args.empty()) args += ",";
                        args += a;
                        match(T_COM);
                    }
                    match(T_RPA);
                    emit("CALL", name, args, "");
                    match(T_SEM);
                }
                break;
            }
            case T_IF: {
                advance();
                match(T_LPA);
                string cond = parseCond();
                match(T_RPA);
                string lElse = newLabel();
                string lEnd = newLabel();
                emit("JZ", cond, "", lElse);
                parseStmt();
                if (match(T_ELSE)) {
                    emit("JMP", "", "", lEnd);
                    emit("LABEL", lElse, "", "");
                    parseStmt();
                    emit("LABEL", lEnd, "", "");
                } else {
                    emit("LABEL", lElse, "", "");
                }
                break;
            }
            case T_WHILE: {
                advance();
                string lStart = newLabel();
                string lEnd = newLabel();
                emit("LABEL", lStart, "", "");
                match(T_LPA);
                string cond = parseCond();
                match(T_RPA);
                emit("JZ", cond, "", lEnd);
                parseStmt();
                emit("JMP", "", "", lStart);
                emit("LABEL", lEnd, "", "");
                break;
            }
            case T_RETURN: {
                advance();
                if (peek().type != T_SEM) {
                    string val = parseExpr();
                    emit("RETURN", val, "", "");
                } else {
                    emit("RETURN", "", "", "");
                }
                match(T_SEM);
                break;
            }
            case T_PRINT: {
                advance();
                string val = parseExpr();
                emit("PRINT", val, "", "");
                match(T_SEM);
                break;
            }
            case T_INPUT: {
                advance();
                if (peek().type == T_ID) {
                    string name = peek().val; advance();
                    emit("INPUT", "", "", name);
                }
                match(T_SEM);
                break;
            }
            case T_LBR: {
                advance();
                while (!match(T_RBR) && peek().type != T_EOF) parseStmt();
                break;
            }
            case T_SEM: advance(); break;
            default: advance(); break;
        }
    }

    string parseExpr() {
        string left = parseTerm();
        while (peek().type==T_PLUS||peek().type==T_MINUS) {
            string op = peek().val; advance();
            string right = parseTerm();
            string tmp = newTmp();
            emit(op, left, right, tmp);
            left = tmp;
        }
        return left;
    }

    string parseTerm() {
        string left = parseFactor();
        while (peek().type==T_MUL||peek().type==T_DIV) {
            string op = peek().val; advance();
            string right = parseFactor();
            string tmp = newTmp();
            emit(op, left, right, tmp);
            left = tmp;
        }
        return left;
    }

    string parseFactor() {
        if (match(T_LPA)) {
            string e = parseExpr();
            match(T_RPA);
            return e;
        }
        if (peek().type == T_MINUS) {
            advance();
            string f = parseFactor();
            string tmp = newTmp();
            emit("NEG", f, "", tmp);
            return tmp;
        }
        if (peek().type==T_NUM||peek().type==T_FLT) {
            string v = peek().val; advance();
            return v;
        }
        if (peek().type == T_ID) {
            string name = peek().val; advance();
            if (match(T_LPA)) {
                // 函数调用表达式
                string args;
                while (peek().type != T_RPA && peek().type != T_EOF) {
                    string a = parseExpr();
                    if (!args.empty()) args += ",";
                    args += a;
                    match(T_COM);
                }
                match(T_RPA);
                string tmp = newTmp();
                emit("CALL", name, args, tmp);
                return tmp;
            }
            return name;
        }
        advance();
        return "err";
    }

    string parseCond() {
        string left = parseExpr();
        if (peek().type==T_LT||peek().type==T_GT||peek().type==T_LE||
            peek().type==T_GE||peek().type==T_EQ||peek().type==T_NE) {
            string op = peek().val; advance();
            string right = parseExpr();
            string tmp = newTmp();
            emit(op, left, right, tmp);
            return tmp;
        }
        return left;
    }

    void printQuads() {
        cout << "========== 四元式中间代码 ==========" << endl;
        cout << "编号\t( 运算符, 操作数1, 操作数2, 结果 )" << endl;
        cout << "----\t---------------------------------" << endl;
        for (const auto& q : quads) {
            printf("%03d\t( %-8s, %-10s, %-10s, %-10s )\n",
                   q.id, q.op.c_str(),
                   q.arg1.c_str(), q.arg2.c_str(), q.result.c_str());
        }
        cout << endl;
        cout << "共 " << quads.size() << " 条四元式" << endl;
    }

    void printSymbolTable() {
        cout << "========== 符号表 ==========" << endl;
        for (const auto& s : syms) {
            cout << "  " << s.name << " : " << s.type;
            if (s.isFunc) cout << " (函数)";
            cout << endl;
        }
        cout << endl;
    }
};

int main(int argc, char* argv[]) {
    SetConsoleOutputCP(65001);
    SetConsoleCP(65001);
    string filename;
    if (argc > 1) filename = argv[1];
    else { cout << "请输入源文件路径: "; cin >> filename; }

    ifstream fin(filename);
    if (!fin.is_open()) {
        cout << "[错误] 无法打开文件: " << filename << endl;
        return 1;
    }
    string source((istreambuf_iterator<char>(fin)), istreambuf_iterator<char>());
    fin.close();

    cout << "========== 源代码 ==========" << endl;
    cout << source << endl << endl;

    // 1. 词法分析
    vector<Tok> tokens = lex(source);
    cout << "========== Token 流 ==========" << endl;
    for (const auto& t : tokens) {
        if (t.type == T_EOF) break;
        cout << "<" << t.val << "> ";
    }
    cout << endl << endl;

    // 2. 语法分析 + 中间代码生成
    IRGen irgen(tokens);
    irgen.parse();

    // 3. 输出
    irgen.printSymbolTable();
    irgen.printQuads();

    return 0;
}

File: compilation-web/js/irgen.js
// Intermediate Code Generator Module (Quadruples)

// Token types (shared with semantic module)
const IR_TK = {
    INT:1,FLOAT:2,VOID:3,IF:4,ELSE:5,WHILE:6,RETURN:7,INPUT:8,PRINT:9,
    ID:10,NUM:11,FLT:12,
    PLUS:13,MINUS:14,MUL:15,DIV:16,
    ASG:17,LT:18,GT:19,LE:20,GE:21,EQ:22,NE:23,AND:24,OR:25,NOT:26,
    LPA:27,RPA:28,LBR:29,RBR:30,LBK:31,RBK:32,SEM:33,COM:34,
    EOF:35,UNK:36
};

// ===== Lexer =====
function irLex(src) {
    const tokens = [];
    const kws = {int:IR_TK.INT,float:IR_TK.FLOAT,void:IR_TK.VOID,if:IR_TK.IF,
        else:IR_TK.ELSE,while:IR_TK.WHILE,return:IR_TK.RETURN,input:IR_TK.INPUT,print:IR_TK.PRINT};
    let p = 0, ln = 1;
    while (p < src.length) {
        while (p < src.length && /\s/.test(src[p])) { if (src[p]==='\n') ln++; p++; }
        if (p >= src.length) break;
        if (p+1 < src.length && src[p]==='/' && src[p+1]==='/') {
            while (p < src.length && src[p]!=='\n') p++;
            continue;
        }
        let c = src[p];
        if (/[a-zA-Z_]/.test(c)) {
            let w = '';
            while (p < src.length && /[a-zA-Z0-9_]/.test(src[p])) w += src[p++];
            tokens.push({ type: kws[w] || IR_TK.ID, val: w, ln });
            continue;
        }
        if (/\d/.test(c) || (c==='.' && p+1<src.length && /\d/.test(src[p+1]))) {
            let n = '', fl = false;
            if (c==='.') { fl = true; n += src[p++]; }
            while (p<src.length && /\d/.test(src[p])) n += src[p++];
            if (p<src.length && src[p]==='.') { fl=true; n+=src[p++]; while(p<src.length&&/\d/.test(src[p])) n+=src[p++]; }
            if (p<src.length && (src[p]==='e'||src[p]==='E')) { fl=true; n+=src[p++]; if(p<src.length&&(src[p]==='+'||src[p]==='-')) n+=src[p++]; while(p<src.length&&/\d/.test(src[p])) n+=src[p++]; }
            tokens.push({ type: fl ? IR_TK.FLT : IR_TK.NUM, val: n, ln });
            continue;
        }
        if (p+1 < src.length) {
            const tw = src.substring(p, p+2);
            const m = {'<=':IR_TK.LE,'>=':IR_TK.GE,'==':IR_TK.EQ,'!=':IR_TK.NE,'&&':IR_TK.AND,'||':IR_TK.OR};
            if (m[tw] !== undefined) { tokens.push({type:m[tw],val:tw,ln}); p+=2; continue; }
        }
        const m1 = {'+':IR_TK.PLUS,'-':IR_TK.MINUS,'*':IR_TK.MUL,'/':IR_TK.DIV,
            '=':IR_TK.ASG,'<':IR_TK.LT,'>':IR_TK.GT,'!':IR_TK.NOT,
            '(':IR_TK.LPA,')':IR_TK.RPA,'{':IR_TK.LBR,'}':IR_TK.RBR,
            '[':IR_TK.LBK,']':IR_TK.RBK,';':IR_TK.SEM,',':IR_TK.COM};
        tokens.push({ type: m1[c] || IR_TK.UNK, val: c, ln });
        p++;
    }
    tokens.push({ type: IR_TK.EOF, val: '', ln });
    return tokens;
}

// ===== IR Generator =====
function IRGenerator(tokens) {
    let pos = 0;
    const quads = [];
    const syms = [];
    let tmpCnt = 1, labelCnt = 0;

    const peek = () => tokens[pos];
    const advance = () => tokens[pos++];
    const match = t => { if (peek().type===t) { advance(); return true; } return false; };
    const newTmp = () => 't' + (tmpCnt++);
    const newLabel = () => 'L' + (labelCnt++);
    const emit = (op, a1, a2, res) => quads.push({id:quads.length, op, arg1:a1||'', arg2:a2||'', result:res||''});

    function parseDeclOrFunc() {
        const typeStr = peek().val; advance();
        if (peek().type === IR_TK.ID) {
            const name = peek().val; advance();
            if (peek().type === IR_TK.LPA) {
                // Function
                syms.push({name, type:typeStr, isFunc:true});
                emit('FUNC', name, '', '');
                advance(); // (
                while (peek().type===IR_TK.INT||peek().type===IR_TK.FLOAT||peek().type===IR_TK.VOID) {
                    const pt = peek().val; advance();
                    if (peek().type === IR_TK.ID) {
                        const pn = peek().val; advance();
                        syms.push({name:pn, type:pt, isFunc:false});
                        emit('PARAM', pn, pt, '');
                    }
                    match(IR_TK.SEM) || match(IR_TK.COM);
                }
                match(IR_TK.RPA);
                if (match(IR_TK.LBR)) {
                    while (!match(IR_TK.RBR) && peek().type !== IR_TK.EOF) parseStmt();
                }
                emit('ENDFUNC', name, '', '');
                match(IR_TK.SEM);
            } else {
                // Variable
                syms.push({name, type:typeStr, isFunc:false});
                emit('DECL', name, typeStr, '');
                match(IR_TK.SEM);
            }
        }
    }

    function parseStmt() {
        switch (peek().type) {
            case IR_TK.INT: case IR_TK.FLOAT: {
                const typeStr = peek().val; advance();
                if (peek().type === IR_TK.ID) {
                    const name = peek().val; advance();
                    syms.push({name, type:typeStr, isFunc:false});
                    emit('DECL', name, typeStr, '');
                    if (match(IR_TK.ASG)) {
                        const val = parseExpr();
                        emit('=', val, '', name);
                    }
                    match(IR_TK.SEM);
                }
                break;
            }
            case IR_TK.ID: {
                const name = peek().val; advance();
                if (match(IR_TK.ASG)) {
                    const val = parseExpr();
                    emit('=', val, '', name);
                    match(IR_TK.SEM);
                } else if (match(IR_TK.LPA)) {
                    let args = '';
                    while (peek().type !== IR_TK.RPA && peek().type !== IR_TK.EOF) {
                        const a = parseExpr();
                        if (args) args += ',';
                        args += a;
                        match(IR_TK.COM);
                    }
                    match(IR_TK.RPA);
                    emit('CALL', name, args, '');
                    match(IR_TK.SEM);
                }
                break;
            }
            case IR_TK.IF: {
                advance();
                match(IR_TK.LPA);
                const cond = parseCond();
                match(IR_TK.RPA);
                const lElse = newLabel();
                const lEnd = newLabel();
                emit('JZ', cond, '', lElse);
                parseStmt();
                if (match(IR_TK.ELSE)) {
                    emit('JMP', '', '', lEnd);
                    emit('LABEL', lElse, '', '');
                    parseStmt();
                    emit('LABEL', lEnd, '', '');
                } else {
                    emit('LABEL', lElse, '', '');
                }
                break;
            }
            case IR_TK.WHILE: {
                advance();
                const lStart = newLabel();
                const lEnd = newLabel();
                emit('LABEL', lStart, '', '');
                match(IR_TK.LPA);
                const cond = parseCond();
                match(IR_TK.RPA);
                emit('JZ', cond, '', lEnd);
                parseStmt();
                emit('JMP', '', '', lStart);
                emit('LABEL', lEnd, '', '');
                break;
            }
            case IR_TK.RETURN: {
                advance();
                if (peek().type !== IR_TK.SEM) {
                    const val = parseExpr();
                    emit('RETURN', val, '', '');
                } else {
                    emit('RETURN', '', '', '');
                }
                match(IR_TK.SEM);
                break;
            }
            case IR_TK.PRINT: {
                advance();
                const val = parseExpr();
                emit('PRINT', val, '', '');
                match(IR_TK.SEM);
                break;
            }
            case IR_TK.INPUT: {
                advance();
                if (peek().type === IR_TK.ID) {
                    const name = peek().val; advance();
                    emit('INPUT', '', '', name);
                }
                match(IR_TK.SEM);
                break;
            }
            case IR_TK.LBR: {
                advance();
                while (!match(IR_TK.RBR) && peek().type !== IR_TK.EOF) parseStmt();
                break;
            }
            case IR_TK.SEM: advance(); break;
            default: advance(); break;
        }
    }

    function parseExpr() {
        let left = parseTerm();
        while (peek().type===IR_TK.PLUS||peek().type===IR_TK.MINUS) {
            const op = peek().val; advance();
            const right = parseTerm();
            const tmp = newTmp();
            emit(op, left, right, tmp);
            left = tmp;
        }
        return left;
    }

    function parseTerm() {
        let left = parseFactor();
        while (peek().type===IR_TK.MUL||peek().type===IR_TK.DIV) {
            const op = peek().val; advance();
            const right = parseFactor();
            const tmp = newTmp();
            emit(op, left, right, tmp);
            left = tmp;
        }
        return left;
    }

    function parseFactor() {
        if (match(IR_TK.LPA)) {
            const e = parseExpr();
            match(IR_TK.RPA);
            return e;
        }
        if (peek().type === IR_TK.MINUS) {
            advance();
            const f = parseFactor();
            const tmp = newTmp();
            emit('NEG', f, '', tmp);
            return tmp;
        }
        if (peek().type===IR_TK.NUM||peek().type===IR_TK.FLT) {
            const v = peek().val; advance();
            return v;
        }
        if (peek().type === IR_TK.ID) {
            const name = peek().val; advance();
            if (match(IR_TK.LPA)) {
                let args = '';
                while (peek().type !== IR_TK.RPA && peek().type !== IR_TK.EOF) {
                    const a = parseExpr();
                    if (args) args += ',';
                    args += a;
                    match(IR_TK.COM);
                }
                match(IR_TK.RPA);
                const tmp = newTmp();
                emit('CALL', name, args, tmp);
                return tmp;
            }
            return name;
        }
        advance();
        return 'err';
    }

    function parseCond() {
        const left = parseExpr();
        const t = peek().type;
        if (t===IR_TK.LT||t===IR_TK.GT||t===IR_TK.LE||t===IR_TK.GE||t===IR_TK.EQ||t===IR_TK.NE) {
            const op = peek().val; advance();
            const right = parseExpr();
            const tmp = newTmp();
            emit(op, left, right, tmp);
            return tmp;
        }
        return left;
    }

    function parse() {
        while (peek().type !== IR_TK.EOF) {
            if (peek().type===IR_TK.INT||peek().type===IR_TK.FLOAT||peek().type===IR_TK.VOID) {
                parseDeclOrFunc();
            } else {
                parseStmt();
            }
        }
    }

    return { parse, quads: () => quads, syms: () => syms };
}

// ===== Main Entry =====
function runIRGen() {
    const src = document.getElementById('irgen-input').value;
    const tokens = irLex(src);
    const gen = IRGenerator(tokens);
    gen.parse();

    const quads = gen.quads();
    const syms = gen.syms();

    // Token stream
    let tokHtml = '<div class="token-stream">';
    tokens.forEach(t => {
        if (t.type === IR_TK.EOF) return;
        tokHtml += `<span class="token-tag">&lt;${t.val}&gt;</span> `;
    });
    tokHtml += '</div>';
    document.getElementById('irgen-tokens').innerHTML = tokHtml;

    // Symbol table
    let stHtml = '<table class="parse-table sym-table"><tr><th>名称</th><th>类型</th><th>类别</th></tr>';
    syms.forEach(s => {
        stHtml += `<tr><td>${s.name}</td><td>${s.type}</td><td>${s.isFunc ? '函数' : '变量'}</td></tr>`;
    });
    stHtml += '</table>';
    document.getElementById('irgen-symtable').innerHTML = stHtml;

    // Quadruples
    let qHtml = '<div class="quad-header"><code>编号 &nbsp; ( 运算符, 操作数1, 操作数2, 结果 )</code></div>';
    qHtml += '<table class="parse-table quad-table"><tr><th>#</th><th>运算符</th><th>操作数1</th><th>操作数2</th><th>结果</th></tr>';
    quads.forEach(q => {
        let cls = '';
        if (q.op==='LABEL') cls = 'quad-label';
        else if (q.op==='JZ'||q.op==='JMP') cls = 'quad-jump';
        else if (q.op==='FUNC'||q.op==='ENDFUNC') cls = 'quad-func';
        else if (q.op==='RETURN'||q.op==='PRINT'||q.op==='INPUT'||q.op==='DECL'||q.op==='PARAM'||q.op==='CALL') cls = 'quad-special';
        qHtml += `<tr class="${cls}"><td>${String(q.id).padStart(3,'0')}</td><td><strong>${q.op}</strong></td><td>${q.arg1}</td><td>${q.arg2}</td><td>${q.result}</td></tr>`;
    });
    qHtml += '</table>';
    qHtml += `<div class="mt-2 text-gray-500 text-xs">共 ${quads.length} 条四元式</div>`;
    document.getElementById('irgen-quads').innerHTML = qHtml;
}
